### PR TITLE
usm: change system-probe and proces-agent communication 

### PR DESCRIPTION
**Pull request by amitslavin**
### What does this PR do?

This PR adds support for a gRPC server using a Unix socket in the system-probe module. It also includes the introduction of a GetConnections function, which is utilized on the process agent's side.

### Motivation

Change system probe and process agent communication to improve system probe performance 

### Describe how to test/QA your changes

In order to enable feature you need to add the current flags to system-probe and process-agent:
`use_grpc` and `grpc_socket_file_path` (if you do not want to use the default sock file) 

Then just create some traffic in order to send real connections.

Validate that the connections was received successfully  

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: cmd/system-probe/modules/network_tracer.go [hunk 1]
```diff
@@ -78,13 +80,83 @@ type networkTracer struct {
...
+	connectionserver.UnsafeSystemProbeServer
+	maxConnsPerMessage int
 }
 
 func (nt *networkTracer) GetStats() map[string]interface{} {
 	stats, _ := nt.tracer.GetStats()
 	return stats
 }
 
+func getConnectionsFromMarshaler(marshaler encoding.Marshaler, cs *network.Connections) ([]byte, error) {
```
Comment (guyarb) [inline cmd/system-probe/modules/network_tracer.go:93]: document all methods here

File: cmd/system-probe/modules/network_tracer.go [hunk 2]
```diff
@@ -78,13 +80,83 @@ type networkTracer struct {
...
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func (nt *networkTracer) GetConnections(req *connectionserver.GetConnectionsRequest, s2 connectionserver.SystemProbe_GetConnectionsServer) error {
+	start := time.Now()
+
+	runCounter := atomic.NewUint64(0)
```
Comment (guyarb) [inline cmd/system-probe/modules/network_tracer.go:113]: this is a "bug" and does not match to the http version

File: cmd/system-probe/modules/process.go [hunk 3]
```diff
@@ -9,13 +9,15 @@ package modules
 
 import (
 	"errors"
+
```
Comment (guyarb) [inline cmd/system-probe/modules/process.go:12]: remove

File: pkg/config/config_windows.go [hunk 4]
```diff
@@ -25,6 +25,7 @@ var (
 
 	// defaultSystemProbeAddress is the default address to be used for connecting to the system probe
 	defaultSystemProbeAddress     = "localhost:3333"
+	defaultGRPCSystemProbeAddress = ""
```
Comment (guyarb) [inline pkg/config/config_windows.go:28]: we probably need to add code to ignore grpc if windows is detected

File: pkg/process/net/common.go [hunk 5]
```diff
@@ -159,6 +161,36 @@ func (r *RemoteSysProbeUtil) GetConnections(clientID string) (*model.Connections
 	return conns, nil
 }
 
+// GetConnectionsGRPC returns a set of active network connections, retrieved from the system probe grpc server
+func (r *RemoteSysProbeUtil) GetConnectionsGRPC(clientID, unixPath string) (*model.Connections, error) {
```
Comment (guyarb) [inline pkg/process/net/common.go:165]: do we want to wrap the errors with indicate message?

File: pkg/process/net/common.go [hunk 6]
```diff
@@ -159,6 +161,36 @@ func (r *RemoteSysProbeUtil) GetConnections(clientID string) (*model.Connections
...
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := response.Recv()
+	if err != nil {
+		return nil, err
+	}
+
+	conns, err := netEncoding.GetUnmarshaler("application/protobuf").Unmarshal(res.Data)
```
Comment (guyarb) [inline pkg/process/net/common.go:187]: can't we use the const that already exists for `application/protobuf`?

Comment (guyarb) [inline pkg/process/net/common.go:187]: i think we can simply go with
```suggestion
	return netEncoding.GetUnmarshaler("application/protobuf").Unmarshal(res.Data)
```

**Review by guyarb — CHANGES_REQUESTED** [review]

**Comment by guyarb** [comment]
Add QA steps

File: cmd/system-probe/modules/network_tracer.go (hunk 2 above)
Comment (amitslavin) [reply cmd/system-probe/modules/network_tracer.go:113]: fixed? or I miss understood it

**Pull request closed without merging**